pennylane/interfaces/execution.py

```diff
@@ -479,17 +482,50 @@ def cost_fn(params, x):
     else:
         batch_execute = set_shots(device, override_shots)(device.batch_execute)
 
-    expand_fn = _preprocess_expand_fn(expand_fn, device, max_expansion)
+    if transform_program is None:
+        expand_fn = _preprocess_expand_fn(expand_fn, device, max_expansion)
 
-    #### Executing the configured setup #####
+        #### Executing the configured setup #####
 
-    tapes, batch_fn, config = _batch_transform(
-        tapes, device, config, override_shots, device_batch_transform
-    )
+        tapes, batch_fn, config = _batch_transform(
+            tapes, device, config, override_shots, device_batch_transform
+        )
+    else:
+        # TODO: add the device preprocessing and expand fn to the program
+        # transform_program.add_preprocessing()
+        expand_fn = None
+
+    # Apply all transforms (device pre-processing, compilation)
+    if transform_program is not None and not transform_program.is_empty():

```

Comment:
Instead of having `if None` checks everywhere, we could simply add an initialization line:
```
transform_program = transform_program or TransformProgram()
```

and then have default behaviour for an empty transform program where is simply returns its input and a null postprocessing functions.